maas/resource_maas_network_interface_tag_test.go

```diff
@@ -0,0 +1,161 @@
+					rs, ok := s.RootModule().Resources["maas_network_interface_tag.test"]
+					if !ok {
+						return "", fmt.Errorf("resource not found: %s", "maas_network_interface_tag.test")
+					}
+
+					if rs.Primary.ID == "" {
+						return "", fmt.Errorf("resource id not set")
+					}
+					return fmt.Sprintf("%s:%s", rs.Primary.Attributes["device"], rs.Primary.Attributes["interface_id"]), nil
+				},
+			},
+		},
+	})
+}
+
+func testAccMaasNetworkInterfaceTagConfig(hostname, macAddress string, tagNames ...string) string {
+	return fmt.Sprintf(`
+resource "maas_device" "test" {
+  hostname = %q
+  network_interfaces {
+    mac_address = %q
+  }
+}
+
+resource "maas_network_interface_tag" "test" {
+  device       = maas_device.test.id
+  interface_id = [for iface in maas_device.test.network_interfaces : iface.id if iface.mac_address == %q][0]
+  tags         = %s
+}
+	`, hostname, macAddress, macAddress, fmt.Sprintf("[\"%s\"]", strings.Join(tagNames, "\", \"")))
+}
+
+func testAccCheckMaasNetworkInterfaceTagExists(resourceName string, tagNames ...string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[resourceName]
+		if !ok {
+			return fmt.Errorf("Not found: %s", resourceName)
+		}
+		// Get the system and interface ID from the state ID
+		systemId, interfaceId, err := maas.SplitTagStateId(rs.Primary.ID)
+		if err != nil {
+			return err
+		}
+
+		// Get the existing interface
+		conn := testutils.TestAccProvider.Meta().(*maas.ClientConfig).Client
+		response, err := conn.NetworkInterface.Get(systemId, interfaceId)
+		if err != nil {
+			return err
+		}
+		if response == nil {
+			return fmt.Errorf("MAAS Network Interface (%s) not found.", rs.Primary.ID)
+		}
+
+		// Check the tags exist
+		for _, tag := range tagNames {
+			if !slices.Contains(response.Tags, tag) {
+				return fmt.Errorf("MAAS Network Interface (%s) tags (%s) do not match expected tags: %s", rs.Primary.ID, response.Tags, tagNames)
+			}
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckMaasNetworkInterfaceDestroy(s *terraform.State) error {
+	conn := testutils.TestAccProvider.Meta().(*maas.ClientConfig).Client
+
+	// loop through the resources in state, verifying each maas_network_interface_tag is destroyed
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "maas_network_interface_tag" {
+			continue
+		}
+
+		// Retrieve the system and interface ID from the state ID
+		systemId, interfaceId, err := maas.SplitTagStateId(rs.Primary.ID)
+		if err != nil {
+			return err
+		}
+
+		// Check the interface doesn't exist
+		response, err := conn.NetworkInterface.Get(systemId, interfaceId)
+		if err == nil {
+			if response != nil && response.ID == interfaceId {
+				return fmt.Errorf("MAAS Network Interface (%s) still exists.", rs.Primary.ID)
+			}
+		}
+		// If the error is a 404, the interface is destroyed as expected
+		if !strings.Contains(err.Error(), "404 Not Found") {
+			return err
```

Comment:
I was about to ask if this should be returning the error if this is expected behaviour before I noticed the `!`. Should the comment be changed slightly so it matches the same logical flow as the program?

ie: `If the error is *not* a 404, the interface has not been destroyed as expected`